lib/mock/mock-agent.js

```diff
@@ -0,0 +1,185 @@
+'use strict'
+
+const { kAgentOpts, kAgentCache, kUrl } = require('../core/symbols')
+const { Agent, setGlobalAgent } = require('../agent')
+const Client = require('../core/client')
+const Pool = require('../pool')
+const {
+  kMockAgentSet,
+  kMockAgentGet,
+  kDispatch,
+  kBuildMockDispatch,
+  kClose,
+  kOriginalClientDispatch,
+  kOriginalPoolDispatch,
+  kOriginalClose,
+  kIsClient,
+  kDispatches,
+  kCreatePool,
+  kIsMockEnabled,
+  kIsMockActive,
+  kNetConnect,
+  kGetNetConnect
+} = require('./mock-symbols')
+const MockClient = require('./mock-client')
+const MockPool = require('./mock-pool')
+const { matchValue, checkNetConnect } = require('./mock-utils')
+const { InvalidArgumentError } = require('../core/errors')
+
+class MockAgent extends Agent {
+  constructor (opts) {
+    super(opts)
+    const mockAgent = this
+
+    this[kOriginalClientDispatch] = Client.prototype.dispatch
+    this[kOriginalPoolDispatch] = Pool.prototype.dispatch
+    this[kNetConnect] = true
+    this[kIsMockActive] = true
+    this[kOriginalClose] = this.close
+    this.close = this[kClose]
+
+    if (this[kIsMockEnabled]()) {
+      Client.prototype.dispatch = this[kBuildMockDispatch](this[kOriginalClientDispatch])
+      Pool.prototype.dispatch = this[kBuildMockDispatch](this[kOriginalPoolDispatch])
+    }
+
+    setGlobalAgent(mockAgent)
+  }
+
+  get (origin) {
+    let pool = this[kMockAgentGet](origin)
+    if (!pool) {
+      pool = this[kCreatePool](origin)
+      this[kMockAgentSet](origin, pool)
+    }
+
+    const originalDispatch = this[kIsClient]() ? this[kOriginalClientDispatch] : this[kOriginalPoolDispatch]
+    const isMockEnabled = this[kIsMockEnabled].bind(this)
+    const getNetConnect = this[kGetNetConnect].bind(this)
+
+    pool.dispatch = function dispatch (opts, handler) {
+      if (isMockEnabled()) {
+        let foundMockDispatch = false
+        foundMockDispatch = pool[kDispatch](opts, handler)
+        if (!foundMockDispatch) {
+          const netConnect = getNetConnect()
+          if (checkNetConnect(netConnect, origin)) {
+            originalDispatch.call(this, opts, handler)
+          } else {
+            throw new Error(`Unable to find mock dispatch and real dispatches are disabled for ${origin}`)
+          }
+        }
+      } else {
+        originalDispatch.call(this, opts, handler)
+      }
+    }
+    return pool
+  }
+
+  deactivate () {
+    this[kIsMockActive] = false
+    Client.prototype.dispatch = this[kOriginalClientDispatch]
+    Pool.prototype.dispatch = this[kOriginalPoolDispatch]
```

Comment:
Do not alter module-wide prototypes.

Reply:
Are you referring to all occurrences of this or just this function? If all occurrences, I think this will affect our ability to intercept in the following way:

```js
// In test code
const mockAgent = new MockAgent({ connections: 1 })
const mockClient = mockAgent.get(baseUrl)
mockClient.intercept({
  path: '/foo',
  method: 'GET'
})

// In source code
const client = new Client(baseUrl)
const { statusCode, headers, trailers, body } = await client.request({
  path: '/foo',
  method: 'GET'
})
```

I can't think of an additional way to intercept `Client.prototype.request` (as it does not interact with an agent as far as I understand!) without explicitly overriding `client` with `mockClient` in the test? E.g. when explicitly overriding, it becomes:

```js
// In test code
const mockAgent = new MockAgent({ connections: 1 })
const mockClient = mockAgent.get(baseUrl)
mockClient.intercept({
  path: '/foo',
  method: 'GET'
})

// In source code
let client = new Client(baseUrl)

// Back in test code
client = mockClient // Through retrieving the client or intercepting it's instantiation

// Back in source code
const { statusCode, headers, trailers, body } = await client.request({
  path: '/foo',
  method: 'GET'
})
```

I don't know if you have any additional ideas/thoughts on overriding the dispatch in this situation?

Reply:
I don't think we should intercept things globally - that's exactly my reasoning from the beginning. We can achieve 

```js
import { request, setGlobalAgent, MockAgent } from 'undici'

// In the test
const mockAgent = new MockAgent({ connections: 1 })
setGlobalAgent(mockAgent)
const mockClient = mockAgent.get(baseUrl)
mockClient.intercept({
  path: '/foo',
  method: 'GET'
})

// in the actual code, this is mocked

const {
  statusCode,
  headers,
  trailers,
  body
} = await request('http://localhost:3000/foo')

console.log('response received', statusCode)
console.log('headers', headers)

for await (const data of body) {
  console.log('data', data)
}

console.log('trailers', trailers)
```

There is nothing shared globally minus the "global agent". If somebody wants something more granular, they can pass around the MockClient or use sinon/proxyquire to wrap.

Reply:
My misunderstanding about intercepting globally, apologies - I assumed we still wanted some sort of support for `Client.prototype.request` in addition to anything passing through the agent! Makes perfect sense! I will change and remove this! :)